File: api/src/R/responses.R
#' Model node to probability table names map
#'
#' each node in the network has fixed outcomes which are attributed
#' probabilities. This object allows a map between the node names
#' and the names of the probability tables in the network object.
.model_prob_names = list(
  Technical_Skills = c("Good", "Poor"),
  System_Security = c("Good", "Poor"),
  Checksum = c("Yes", "Self_Generated", "No"),
  Info_Management = c("Sufficient", "Insufficient"),
  Digital_Object = c("Born_digital", "Digitised", "Surrogate"),
  Storage_Medium = c("A", "B", "C"),
  Rep_and_Refresh = c("Good", "Poor"),
  Op_Environment = c("Yes", "No"),
  Physical_Disaster = c("Yes", "No")
)

#' Unpack json
#'
#' Convert unnamed lists in a list of lists to a vector.
#' This ensures that JSON arrays are converted to vectors rather than
#' lists. Operates recursively across the whole list of lists.
#'
#' @param res a list (of lists)
#' @return a list (of lists) where unnamed lists have become vectors.
unpack_json = function(res) {
  purrr::map(res, function(x) {
    if (is.null(names(x))) {
      unlist(x)
    } else {
      unpack_json(x)
    }
  })
}

# @param parsed_json an array of json objects
extract_simple_responses = function(parsed_json) {
  if (is_array(parsed_json)) {
    purrr::map(parsed_json, ~create_simple_responses(.x))
  } else {
    create_simple_responses(parsed_json)
  }
}

is_array = function(parsed_json) {
  is.null(names(parsed_json))
}

advanced_flags = function(parsed_json) {
  if (is_array(parsed_json)) {
    # implies array of json objects
    purrr::map_lgl(parsed_json, ~isTRUE(.x$is_advanced))
  } else {
    isTRUE(parsed_json$is_advanced)
  }
}

format_simple_responses = function(response) {
  res = tibble::tibble(node = names(response), response = response)
  res
}

# @param parsed_json a single json object
create_simple_responses = function(parsed_json) {
  res = list(
    data = format_simple_responses(unpack_json(parsed_json$response)),
    model_name = parsed_json$model_name,
    scenario = parsed_json$scenario,
    notes = parsed_json$notes,
    intellectual_control = parsed_json$intellectual_control,
    renderability = parsed_json$renderability
  )
  class(res) = c("simple_responses", class(res))
  res
}

extract_responses = function(parsed_json) {
  flag = advanced_flags(parsed_json)
  if (is_array(parsed_json)) {
    c(
      extract_simple_responses(parsed_json[!flag]),
      extract_advanced_responses(parsed_json[flag])
    )
  } else {
    if (flag) {
      extract_advanced_responses(parsed_json)
    } else {
      extract_simple_responses(parsed_json)
    }
  }
}

extract_advanced_responses = function(parsed_json) {
  if (is_array(parsed_json)) {
    purrr::map(parsed_json, format_advanced_responses)
  } else {
    format_advanced_responses(parsed_json)
  }
}

# a full advanced model is composed of both the simple
# and advanced parts
format_advanced_responses = function(parsed_json) {
  simple = unpack_json(parsed_json$response)
  advanced = parsed_json$advanced
  overlapping_keys = intersect(names(simple), names(advanced))

  simple_part = simple[setdiff(names(simple), overlapping_keys)]
  simple_part = format_simple_responses(simple_part)
  class(simple_part) = unique(c("simple_responses", class(simple_part)))

  advanced_tables = purrr::map(unpack_json(advanced), ~{
    tibble::as_tibble(.x)
  })

  res = list(
    data = list(
      "simple" = simple_part,
      "advanced" = advanced_tables
    ),
    model_name = parsed_json$model_name,
    scenario = parsed_json$scenario,
    notes = parsed_json$notes,
    intellectual_control = parsed_json$intellectual_control,
    renderability = parsed_json$renderability
  )
  class(res) = c("advanced_responses", class(res))
  res
}


response_type = function(obj) {
  if (inherits(obj, "advanced_responses")) return("advanced")
  else if (inherits(obj, "simple_responses")) return("simple")
  else stop("object does not look like a response.")
}

File: api/src_python/responses.py
"""
Parse and classify incoming JSON request bodies into structured response objects.
"""

from nodes import USER_NODES, REVERSE_NODE_MAP


def _is_array(parsed_json) -> bool:
    """Return True if parsed_json is a JSON array (list without string keys)."""
    return isinstance(parsed_json, list)


def _unpack_json(obj):
    """
    Recursively convert unnamed lists (JSON arrays) to plain Python lists,
    keeping named dicts as dicts.  This mirrors R's unpack_json behaviour.
    """
    if isinstance(obj, dict):
        return {k: _unpack_json(v) for k, v in obj.items()}
    if isinstance(obj, list):
        return [_unpack_json(v) for v in obj]
    return obj


def _is_advanced_flag(obj) -> bool:
    return bool(obj.get("is_advanced", False))


def advanced_flags(parsed_json) -> list:
    """Return a list of booleans indicating which objects are advanced models."""
    if _is_array(parsed_json):
        return [_is_advanced_flag(o) for o in parsed_json]
    return [_is_advanced_flag(parsed_json)]


def _create_simple_responses(obj: dict) -> dict:
    """Build a simple_responses object from a single parsed JSON object."""
    response = _unpack_json(obj.get("response", {}))
    return {
        "type": "simple_responses",
        "data": {"response": response},
        "model_name": obj.get("model_name") or obj.get("modelname"),
        "scenario": obj.get("scenario", ""),
        "notes": obj.get("notes", ""),
        "intellectual_control": obj.get("intellectual_control"),
        "renderability": obj.get("renderability"),
    }


def _create_advanced_responses(obj: dict) -> dict:
    """Build an advanced_responses object from a single parsed JSON object."""
    response = _unpack_json(obj.get("response", {}))
    advanced = _unpack_json(obj.get("advanced", {})) or {}

    # Nodes that appear in both simple and advanced: advanced wins
    overlapping = set(response.keys()) & set(advanced.keys())
    simple_part = {k: v for k, v in response.items() if k not in overlapping}

    return {
        "type": "advanced_responses",
        "data": {
            "simple": {"response": simple_part},
            "advanced": advanced,
        },
        "model_name": obj.get("model_name") or obj.get("modelname"),
        "scenario": obj.get("scenario", ""),
        "notes": obj.get("notes", ""),
        "intellectual_control": obj.get("intellectual_control"),
        "renderability": obj.get("renderability"),
    }


def _extract_single(obj: dict) -> dict:
    if _is_advanced_flag(obj):
        return _create_advanced_responses(obj)
    return _create_simple_responses(obj)


def extract_responses(parsed_json) -> list:
    """Return a list of response objects (simple or advanced) from parsed JSON."""
    if _is_array(parsed_json):
        return [_extract_single(o) for o in parsed_json]
    return [_extract_single(parsed_json)]


def is_array(parsed_json) -> bool:
    return _is_array(parsed_json)
